### PR TITLE
[Speculative] Add heterogeneous vocabulary support for standalone speculative decoding

### PR DESCRIPTION
**Pull request by txh1873749380**
Enable draft models with different vocabularies in speculative decoding. Implements ICML 2025 paper "Lossless Speculative Decoding for Heterogeneous Vocabularies".

## Motivation

Currently, SGLang's speculative decoding requires the draft model and target model to share the same vocabulary. This limits users to using only matched model pairs (e.g., LLaMA draft for LLaMA target).

This PR implements the TLI (Token-Level Intersection) algorithm from the ICML 2025 Oral paper ["Lossless Speculative Decoding for Heterogeneous Vocabularies"](https://arxiv.org/abs/2502.05202) to remove this constraint.

**Algorithm:**
1. Compute vocabulary intersection between draft and target tokenizers
2. Mask draft logits to restrict sampling to intersection tokens
3. Map draft token IDs to target vocab space during verification

## Modifications

- **New file**: `python/sglang/srt/speculative/vocab_intersection.py`
  - `VocabIntersectionMapper` class for vocabulary intersection management
  - `create_vocab_mapper()` factory function

- **Modified files**:
  - `server_args.py`: Add `--enable-heterogeneous-vocab` flag
  - `standalone_worker.py`: Initialize vocab mapper when enabled
  - `eagle_worker.py`: Apply logits masking in draft forward
  - `eagle_info.py`: Map draft tokens to target vocab in verification
  - `multi_layer_eagle_worker.py`: Add vocab_mapper attribute for consistency

**Usage:**
```bash
python -m sglang.launch_server \
    --model meta-llama/Llama-3.1-8B-Instruct \
    --speculative-algorithm STANDALONE \
    --speculative-draft-model-path Qwen/Qwen2-0.5B \
    --enable-heterogeneous-vocab
```

## Accuracy Tests

N/A - This is a compatibility feature. The TLI algorithm is mathematically lossless (output distribution identical to target model). Feature is opt-in and does not affect existing behavior when disabled.

## Benchmarking and Profiling

N/A - This enables new use cases (heterogeneous vocab pairs) rather than optimizing existing ones. Performance depends on vocabulary intersection size between chosen model pairs.

## Checklist

- [x] Format your code according to the Format code with pre-commit.
- [ ] Add unit tests according to the Run and add unit tests.
- [ ] Update documentation according to Write documentations.
- [x] Follow the SGLang code style guidance.

## References

- Paper: [Lossless Speculative Decoding for Heterogeneous Vocabularies](https://arxiv.org/abs/2502.05202) (ICML 2025 Oral)
- HuggingFace: [Universal Assisted Generation](https://huggingface.co/blog/universal_assisted_generation) (Transformers 4.50.0)

### CODE REVIEW

**Comment by txh1873749380** [comment]
Hi @hnyls2002, could you take a look at this PR when you have a chance? It adds heterogeneous vocab support for speculative decoding based on the ICML 2025 TLI paper. Happy to address any feedback. Thanks!
